include/RAJA/policy/hip/MemUtils_HIP.hpp

```diff
@@ -272,10 +293,64 @@ bool setupReducers() { return detail::tl_status.setup_reducers; }
 RAJA_INLINE
 hip_dim_t currentGridDim() { return detail::tl_status.gridDim; }
 
+//! get grid size of current launch
+RAJA_INLINE
+hip_dim_member_t currentGridSize() { return detail::tl_status.gridDim.x *
+                                            detail::tl_status.gridDim.y *
+                                            detail::tl_status.gridDim.z; }
+
 //! get blockDim of current launch
 RAJA_INLINE
 hip_dim_t currentBlockDim() { return detail::tl_status.blockDim; }
 
+//! get block size of current launch
+RAJA_INLINE
+hip_dim_member_t currentBlockSize() { return detail::tl_status.blockDim.x *
+                                             detail::tl_status.blockDim.y *
+                                             detail::tl_status.blockDim.z; }
+
+//! get dynamic shared memory usage for current launch
+RAJA_INLINE
+size_t currentDynamicShmem() { return *detail::tl_status.dynamic_smem; }
+
+//! get maximum dynamic shared memory for current launch
+RAJA_INLINE
+size_t maxDynamicShmem() { return hip::device_prop().sharedMemPerBlock; }
+
+constexpr size_t dynamic_smem_allocation_failure = std::numeric_limits<size_t>::max();
+
+//! Allocate dynamic shared memory for current launch
+//
+//  The first argument is a functional object that takes the maximum number of
```

Comment:
what happens if the number exceeds the number that can fit in shared memory? Is there a fall-back to global memory, for example?  The user docs give the impression that there is no limit. It seems like it would be good to note what happens (e.g., runtime performance) is a user jams a large number of multi reductions into a kernel.

Reply:
That is mentioned in the Returns part of the comment below, but it is unclear on what it actually does to the amount of shmem used by the kernel, I'll update this to be more clear.
At this level there is no fallback. The calling code must check the return value and decide how to fallback to another algorithm or throw an error if allocation fails.

Reply:
Looking at this reminds me that a kernel already using static shared memory has a lower max dynamic shmem which this coding does not cover. I'll have to look into if we can get a better max shmem.

Reply:
It may not matter for our use cases, but it may matter for others.

Reply:
I changed this to always check the actual max dynamic shmem for the kernel, this should help us avoid random unexplained failures.